src/Microsoft.AspNet.PipelineCore/FormFeature.cs

```diff
@@ -1,71 +1,195 @@
 // Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.AspNet.FeatureModel;
 using Microsoft.AspNet.Http;
-using Microsoft.AspNet.HttpFeature;
-using Microsoft.AspNet.PipelineCore.Infrastructure;
+using Microsoft.AspNet.PipelineCore.Collections;
 using Microsoft.AspNet.WebUtilities;
-using Microsoft.AspNet.WebUtilities.Collections;
 
 namespace Microsoft.AspNet.PipelineCore
 {
     public class FormFeature : IFormFeature
     {
-        private readonly IFeatureCollection _features;
-        private readonly FeatureReference<IHttpRequestFeature> _request = FeatureReference<IHttpRequestFeature>.Default;
-        private Stream _bodyStream;
-        private IReadableStringCollection _form;
+        private readonly HttpRequest _request;
 
-        public FormFeature([NotNull] IDictionary<string, string[]> form)
-            : this (new ReadableStringCollection(form))
+        public FormFeature([NotNull] IFormCollection form)
         {
+            Form = form;
         }
 
-        public FormFeature([NotNull] IReadableStringCollection form)
+        public FormFeature([NotNull] HttpRequest request)
         {
-            _form = form;
+            _request = request;
         }
 
-        public FormFeature([NotNull] IFeatureCollection features)
+        public bool HasFormContentType
         {
-            _features = features;
+            get
+            {
+                // Set directly
+                if (Form != null)
+                {
+                    return true;
+                }
+
+                return HasApplicationFormContentType() || HasMultipartFormContentType();
+            }
         }
 
-        public async Task<IReadableStringCollection> GetFormAsync(CancellationToken cancellationToken)
+        public IFormCollection Form { get; set; }
```

Comment:
`Form` is the storage "field", we just don't put fields on interfaces. We want the raw value to be directly accessible so you can do things like check if the form has been read yet without actually triggering a read.

The only reason I put a Form property on HttpRequest instead of a ReadForm() method is to make it easier to assign the form in a unit test.  ReadForm() would better convey the potential costs of accessing the form.